Plugins/Events/Events/DMActionEvents.cpp

```diff
@@ -1,36 +1,198 @@
 #include "Events/DMActionEvents.hpp"
+#include "API/Functions.hpp"
+#include "API/CNWSPlayer.hpp"
 #include "API/CNWSCreature.hpp"
 #include "API/CNWSCreatureStats.hpp"
+#include "API/CNWSMessage.hpp"
+#include "API/Types.hpp"
 #include "API/Version.hpp"
+#include "API/Constants.hpp"
 #include "Events.hpp"
 #include "Services/Patching/Patching.hpp"
-
+#include "Helpers.hpp"
 namespace Events {
 
 using namespace NWNXLib;
 using namespace NWNXLib::API;
 using namespace NWNXLib::Platform;
 
-DMActionEvents::DMActionEvents(NWNXLib::ViewPtr<NWNXLib::Services::PatchingProxy> patching)
+DMActionEvents::DMActionEvents(NWNXLib::ViewPtr<NWNXLib::Services::HooksProxy> hooker)
 {
-    patching->PatchWithCall(0x0818C825, 0x0818C82A - 0x0818C825, &SetExperiencePatch); NWNX_EXPECT_VERSION(8109);
-    patching->PatchWithCall(0x0818C8E2, 0x0818C8E7 - 0x0818C8E2, &AddGoldPatch); NWNX_EXPECT_VERSION(8109);
+    hooker->RequestSharedHook<Functions::CNWSMessage__HandlePlayerToServerDungeonMasterMessage, int32_t,
+        CNWSMessage*, CNWSPlayer*, uint8_t, int32_t>(&HandleDMMessageHook);
 }
-
-void DMActionEvents::SetExperiencePatch(CNWSCreatureStats* stats, uint32_t exp, int32_t doLevel)
+template <typename T>
+static T& PeekMessage(CNWSMessage *pMessage, int32_t offset)
 {
-    const uint32_t oldExp = stats->m_nExperience;
-    stats->SetExperience(exp, doLevel);
-    Events::PushEventData("OLD_EXP", std::to_string(oldExp));
-    Events::PushEventData("NEW_EXP", std::to_string(exp));
-    Events::SignalEvent("NWNX_ON_DM_SET_EXP", stats->m_pBaseCreature->m_idSelf);
+    uint8_t *ptr = pMessage->m_pnReadBuffer + pMessage->m_nReadBufferPtr + offset;
+    return *(T*)ptr;
```

Comment:
This is going to break strict aliasing. Maybe return T instead of T& and do

```cpp
static_assert(std::is_pod<T>::value);
T value;
std::memcpy(&value, ptr, sizeof(T));
return T;
```

Reply:
Strict aliasing allows casting to/from `char*` (and related types, such as `uint8_t*`). But I totally did forget about it (I always compile my code with `-fno-strict-aliasing`), so definitely a good idea to change as you suggested, otherwise I'll just forget again and do something more wrong.

I went with `T&` so the hooks could in theory modify the message, but that should be done by exclusive hooks only, and there's better ways to do it.

Reply:
I think you can freely cast anything *to* a char but you can't freely cast to anything *from* a char? Something to do with alignment ... ? I can never remember the exact rules and it's too early to read standards. I try to be on the safe side and never cast to anything from a char, but I could just be a little bit mixed up in my understanding!

Reply:
Yeah, that sounds about right. Fixed.